src/main/js/utils/generate.js

```diff
@@ -6,57 +6,109 @@ import {
+    .then(response => response.json());
+};
+
+const redirectToGenerator = (schedules, search = '') => startRedirect({
+  pathname: '/generator',
+  hash: schedules.length > 0 ? '#schedules' : '#timetable',
+  search: search.replace('==', '='),
+});
+
+const generateSchedules = (values, loadQuery = '') => (dispatch, getState) => {
   const { courseList } = getState().courses;
   if (courseList.length !== 0) {
     dispatch(startGenerating());
+
+    const shortCode = loadQuery || !values.genURL ? Promise.resolve(loadQuery.substring(2))
+      : fetchShortUrl(courseList, values);
+
     const filteredResults = filterCourses(values, courseList);
     dispatch(filteredCourses(filteredResults));
-    const filteredCourseList = filteredResults.map(result => (
-      result.filtered));
-    let schedules = [];
-    if (!filteredCourseList.find(list => list.length === 0)) {
-      const scheduleMaker = new ScheduleMaker(filteredCourseList, parseInt(values.gap, 10));
-      try {
-        schedules = scheduleMaker.makeSchedules();
-      } catch (e) {
-        schedules = e;
-      }
-    }
-    const queryObject = {
-      ...values,
-      c: courseList.map((course) => {
-        const name = `${course[0].subject}${course[0].courseNumber}`;
-        return JSON.stringify({
-          name,
-          ...(values.courses ? values.courses[`${name}${course.id}`] : {}),
-        });
-      }),
-    };
-    delete queryObject.courses;
-    (loadQuery ? Promise.resolve() : fetch(`${shortenUrl}?${stringify({ data: JSON.stringify(queryObject), prefix: 'generator' })}`))
-      .then(response => (loadQuery ? loadQuery.substring(2) : response.json()))
-      .then((result) => {
-        const querystring = `q=${result}`;
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: result ? querystring : '',
-        }));
-      })
-      .catch(() => {
-        dispatch(startRedirect({
-          pathname: '/generator',
-          hash: schedules.length > 0 ? '#schedules' : '#timetable',
-          search: '',
+    let filteredCourseList = filteredResults.map(result => result.filtered);
+
+    const gpasLoading = values.sortByGPA === 'yes' ? addGPA(filteredCourseList) : Promise.reject();
+
+    gpasLoading
+      .then((res) => {
+        dispatch(addGrades(res));
+        filteredCourseList = filteredCourseList.map(list => list.sort((course1, course2) => {
+          const name = `${course1.subject}${course1.courseNumber}`;
+          const instructor1 = getInstructorLastName(course1.instructor); // Backend only holds instructors' last name
+          const instructor2 = getInstructorLastName(course2.instructor);
+
+          /** Can't sort if the backend doesn't hold data for the entire course */
+          if (!res[name]) return 0;
+
+          /** Assume when instructor === 'Staff', then the course has a gpa of 0.00 */
```

Comment:
Subject to change based on my suggestion.

Also, do they ever have Staff listed in the grade distributions? You wouldn't need to check for it specifically then

Reply:
Removed.